src/v/cloud_storage/async_manifest_view.cc

```diff
@@ -1401,62 +1332,47 @@ async_manifest_view::hydrate_manifest(
     }
 }
 
-int32_t async_manifest_view::search_spillover_manifests(
-  async_view_search_query_t query) const {
-    if (!_manifests.has_value()) {
-        throw std::logic_error(fmt_with_ctx(
-          fmt::format,
-          "can't hydrate spillover manifest, bucket has to be scanned first"));
-    }
-    if (
-      _last_stm_start_offset
-      < _stm_manifest.get_start_offset().value_or(model::offset{})) {
-        auto errc = make_error_code(error_outcome::repeat);
-        throw std::system_error(
-          errc,
-          fmt_with_ctx(
-            fmt::format,
-            "list of spillover manifests is outdated, current "
-            "start_offset: {}, cached: {}",
-            _stm_manifest.get_start_offset().value_or(model::offset{}),
-            _last_stm_start_offset));
-    }
-    if (_last_archive_start_offset < _stm_manifest.get_archive_start_offset()) {
-        auto errc = make_error_code(error_outcome::repeat);
-        throw std::system_error(
-          errc,
-          fmt_with_ctx(
-            fmt::format,
-            "list of spillover manifests is outdated, current "
-            "archive_start_offset: {}, cached: {}",
-            _stm_manifest.get_archive_start_offset(),
-            _last_archive_start_offset));
-    }
-
-    // Perform simple scan of the manifest list.
-    vlog(
-      _ctxlog.trace,
-      "Scanning manifests list with {} elements, query: {}",
-      _manifests->components.size(),
-      to_string(query));
-    auto it = std::find_if(
-      _manifests->components.begin(),
-      _manifests->components.end(),
-      [this, query](const spillover_manifest_path_components& c) {
-          vlog(
-            _ctxlog.trace,
-            "Scanning manifest with components {}, query: {}",
-            c,
-            to_string(query));
-          return contains(c, query);
+namespace {
+bool contains(
+  const segment_meta& meta, const async_view_search_query_t& query) {
+    return ss::visit(
+      query,
+      [&](model::offset o) {
+          return o >= meta.base_offset && o <= meta.committed_offset;
+      },
+      [&](kafka::offset k) {
+          return k >= meta.base_kafka_offset() && k < meta.next_kafka_offset();
+      },
+      [&](model::timestamp t) {
+          return t >= meta.base_timestamp && t <= meta.max_timestamp;
       });
+}
+} // namespace
 
-    if (it == _manifests->components.end()) {
-        return -1;
+std::optional<segment_meta> async_manifest_view::search_spillover_manifests(
+  async_view_search_query_t query) const {
+    const auto& manifests = _stm_manifest.get_spillover_map();
+    // TODO: implement late materialization for filtering
+    for (const auto& it : manifests) {
```

Comment:
nit: haven't tried yet, but we should be able now to write something like

`std::ranges::find_if(std::ranges::ref_view(manifest), [&](auto e){return contains(e, query);});
`